ConfigSpace/hyperparameters.pyx

```diff
@@ -2071,6 +2509,60 @@ cdef class CategoricalHyperparameter(Hyperparameter):
                          "OrdinalHyperparameter, but is "
                          "<cdef class 'ConfigSpace.hyperparameters.CategoricalHyperparameter'>")
 
+    def pdf(self, vector: np.ndarray) -> np.ndarray:
+        """
+        Computes the probability density function of the parameter in 
+        the original parameter space (the one specified by the user).
+        For each parameter type, there is also a method _pdf which 
+        operates on the transformed (and possibly normalized) parameter
+        space. Only legal values return a positive probability density,
+        otherwise zero.
+        
+        Parameters
+        ----------
+        vector: np.ndarray
+            the (N, ) vector of imputs for which the probability density
+            function is to be computed.
+        
+        Returns
+        ----------
+        np.ndarray(N, )
+            Probability density values of the input vector
+        """
+        # this check is to ensure shape is right (and np.shape does not work in cython)
+        if vector.ndim != 1:
+            raise ValueError("Method pdf expects a one-dimensional numpy array")
+        vector = np.array(self._inverse_transform(vector))
+        return self._pdf(vector)
+        
+    def _pdf(self, vector: np.ndarray) -> np.ndarray:
+        """
+        Computes the probability density function of the parameter in 
+        the transformed (and possibly normalized, depends on the parameter 
+        type) space. As such, one never has to worry about log-normal 
+        distributions, only normal distributions (as the inverse_transform
+        in the pdf method handles these).
+        
+        Parameters
+        ----------
+        vector: np.ndarray
+            the (N, ) vector of imputs for which the probability density
+            function is to be computed.
+        
+        Returns
+        ----------
+        np.ndarray(N, )
+            Probability density values of the input vector
+        """
+        probs = np.array(self.probabilities)
+        res = np.array(probs[vector.astype(int)])
```

Comment:
So, the vector that gets passed for ['cat', 'dog', 'mouse'] is [0.0, 1.0, 2.0], since each vector gets transformed to its corresponding index (but in float form). To be able to retrieve the element corresponding to the index, the float must be converted to int.

Reply:
So, the _pdf method only ever operates on arrays of floats

Reply:
Cool, thanks for clarifying that. I think at some point we need to go back and document everything a lot more clearly, i.e. how would one implement a hyperparameter and what is the expected input and output of the functions that need to be implemented. Thanks for confirming this :)